src/nvim/tui/tui.c

```diff
@@ -527,7 +529,13 @@ static void tui_flush(UI *ui)
   }
 
   unibi_goto(ui, data->row, data->col);
+
+  if (!data->busy) {
+    unibi_out(ui, unibi_cursor_normal);
+  }
+
   flush_buf(ui);
+  unibi_out(ui, unibi_cursor_invisible);
```

Comment:
In the patch `invis_buf` is not flushed:
```diff
-uv_write(&req, (uv_stream_t *)&data->output_handle, buf + buf_index, 1, NULL);
+uv_write(&req, (uv_stream_t *)&data->output_handle, buf, buf_index + 1, NULL);
```
Is it correct that the following patch would not introduce cursor flashing (not sure if your answer said the snippet would not work)?
Either the cursor was already hidden and no "show cursor" since then or the first command in the flushed buffer is "hide cursor". So a flushed buffer would not be drawn with a shown cursor. I did some simple test and compared this version with your patch (compared debug output), but could not find that your patch used fewer "hide cursor" commands.

```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index 610419d..3f43c56 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -108,6 +108,7 @@ void tui_start(void)
     data->ut = unibi_dummy();
   }
   fix_terminfo(data);
+  unibi_out(ui, unibi_cursor_invisible);
   // Enter alternate screen and clear
   unibi_out(ui, unibi_enter_ca_mode);
   unibi_out(ui, unibi_clear_screen);
@@ -530,12 +531,7 @@ static void tui_flush(UI *ui)
 
   unibi_goto(ui, data->row, data->col);
 
-  if (!data->busy) {
-    unibi_out(ui, unibi_cursor_normal);
-  }
-
   flush_buf(ui);
-  unibi_out(ui, unibi_cursor_invisible);
 }
 
 static void tui_suspend(UI *ui)
@@ -792,11 +788,27 @@ static void flush_buf(UI *ui)
   static uv_write_t req;
   static uv_buf_t buf;
   TUIData *data = ui->data;
+
+  static int i = 0;
+
+  i++;
+  if (!data->busy) {
+    unibi_out(ui, unibi_cursor_normal);
+    fprintf(stderr, "flush %d: drawing -> cursor on -> pending cursor off\n",
+            i);
+  } else {
+    fprintf(stderr, "flush %d: drawing\n", i);
+  }
+
   buf.base = data->buf;
   buf.len = data->bufpos;
   uv_write(&req, (uv_stream_t *)&data->output_handle, &buf, 1, NULL);
   uv_run(data->write_loop, UV_RUN_DEFAULT);
   data->bufpos = 0;
+
+  if (!data->busy) {
+    unibi_out(ui, unibi_cursor_invisible);
+  }
 }
 
 static void destroy_screen(TUIData *data)
```